File: chelseashin/Programmers/02_Stack&Queue/다리를지나는트럭.py
# from 8:20

bridge_length = 2
weight = 10
truck_weight = [7, 4, 5, 6]

def solution(N, W, trucks):
    answer = 0
    bridge = [0] * N
    while trucks:
        if sum(bridge[1:]) + trucks[0] <= W:
            bridge.append(trucks.pop(0))
        else:
            while bridge[1] == 0:
                bridge.pop(0)
                bridge.append(0)
                answer += 1
            bridge.append(0)
        answer += 1
        bridge.pop(0)
        # print(answer, bridge, trucks)
    answer += N
    return answer

print(solution(bridge_length, weight, truck_weight))

File: hyeanwooRoh/programmers/다리를지나는트럭.js
// 52분
// 초반에 문제를 잘못이해해서 어렵게 풀었다
// 알고보니 순서가 정해져 있어서 조건맞추어 구현했다.

const bridge_length = 2;
const weight = 10;
const truck_weights = [7,4,5,6]

function solution(bridge_length, weight, truck_weights) {
  let sec = 1;
  let bridge = new Array(bridge_length).fill(0);
  let onBridgeWeight = 0;
  let truck_cnt = truck_weights.length;
  
  while(truck_cnt != 0) {
    if(onBridgeWeight + truck_weights[0] <= weight) {
      let truck = truck_weights.shift();
      onBridgeWeight += truck;
      bridge[bridge.length-1] = truck;
    }
    if(bridge[0] !== 0) {
      onBridgeWeight -= bridge[0];
      bridge[0] = 0;
      truck_cnt--;
    }
    for(let i = 1; i < bridge.length; i++) {
      if(bridge[i] !== 0) {
        bridge[i-1] = bridge[i];
        bridge[i] = 0;
      }
    }
  
    sec++;
  }

  return sec;
}


console.log(solution(bridge_length, weight, truck_weights));